### DISCUSSION TITLE
Angle Of Swerve Wheels Won't Stabilize On Initial Enabling

### DISCUSSION THREAD

**Discussion by TotinoGracePizzaPis** [BroncBotz3481/YAGSL-Example#141] [Q&A]
Hello.
We are using MK4i, Thrifty Bot Encoders, and the NavX Micro IMU. Whenever we enable the robot, the angle is constantly changing and does not stabilize. Additionally, when we move our joystick, the drive speed goes out of control. We are unsure what to do from here. Here is our code: https://gitlab.com/totino-grace-robotics-control-team/swerve2024_yagsl

**Comment by thenetworkgrinch** [comment]
Have you read this?

https://yagsl.gitbook.io/yagsl/configuring-yagsl/when-to-invert

**Reply by TotinoGracePizzaPis** [reply]
When we rotated the wheels counterclockwise, the absolute encoder and angle did increase. We tried setting the inverted angle from true to false in our .json files and that still didn't work.

**Reply by thenetworkgrinch** [reply]
The absolute encoder will not be reflected as increasing just yet. However it should be set to true in your case. Maybe lower your PID while you're at it too

**Reply by TotinoGracePizzaPis** [reply]
What are you referring you say "It should be set to true"?
Also, our absolute encoder offset value changes every time we enable, disable, and line up the wheels again. For example, when we set the offset to 7.6 after we run it and try setting it again, it'll change to a value like 244.6
Our set point angle is also switching from -45 to 135 on our SmartDashboard.

**Reply by thenetworkgrinch** [reply]
I am referring to "absoluteEncoderInverted"

**Reply by TotinoGracePizzaPis** [reply]
Those are all set to true. None of them should be inverted since it didn't follow the rules. 
What should we do next?

**Reply by thenetworkgrinch** [reply]
Are your thriftybot readings good? Do they change rapidly or not chabge?

**Reply by TotinoGracePizzaPis** [reply]
When we run the robot, the encoder numbers increase and wrap around 360 degrees. The values do change rapidly.

**Reply by thenetworkgrinch** [reply]
You may have a bad range or connection to the absolute encoders. I have also heard you need to tune your absolute encoder offset and conversion factors to solve this.

**Reply by thenetworkgrinch** [reply]
Have you read this?

https://www.chiefdelphi.com/t/fluctuating-readings-from-thriftybot-absolute-encoders/450286

**Reply by TotinoGracePizzaPis** [reply]
Hello! We have switched from ThriftyBots to CANCoders. The coders are flashing green, so we know they are working. 

We have set our offsets already. However, our wheels are not fully turning when we want them to. If we tell them to turn left, it'll ping pong left and right. Our conversion factors also seem to be correct. What should we do?

**Reply by thenetworkgrinch** [reply]
Invert the angle motor or absolute encoder to turn CCW+ one is ojt of phase

**Reply by TotinoGracePizzaPis** [reply]
They are all turning counter clockwise and increasing with this turn. We followed a wiki page and it said we didn't need to invert any motors.

For reference, this is our code: https://gitlab.com/totino-grace-robotics-control-team/2024_robot_code

**Reply by thenetworkgrinch** [reply]
Were they CCW+ from a top down view?

**Reply by TotinoGracePizzaPis** [reply]
Yes. We positioned the robot so we were facing the back and pointed the bevels to our left, which would be pointing to the Front Left and Back Left. We turned the wheels manually counterclockwise according to a top down view.

**Reply by thenetworkgrinch** [reply]
How about the absolute encoder?

**Reply by TotinoGracePizzaPis** [reply]
The absolute encoder value increases when we turn CCW. However, we've noticed that when our encoder offset value changes each time we realign the wheels to 0.

**Reply by thenetworkgrinch** [reply]
Then your magnets may be moving inside the modules.

**Reply by tfield** [reply]
Agreed - we had this problem with our cancoders - one of them came unglued and it would rotate a few degrees from time to time.  While translating in a straight line without rotating it was plain to see that the wheels weren't aligned straight anymore.  When we took the cancoder apart the magnet slid out as if it hadn't been glued at all.  Easy fix though.
